File: benchmark/cpp/benchmark.cc
#include <chrono>
#include <cmath>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>

#include "hilbert_curve_sort.h"

auto calc_dist(const std::array<double, 2> &point1,
               const std::array<double, 2> &point2) -> double {
  const double x_1 = point1[0];
  const double y_1 = point1[1];
  const double x_2 = point2[0];
  const double y_2 = point2[1];

  return ceil(sqrt(((x_1 - x_2) * (x_1 - x_2)) + ((y_1 - y_2) * (y_1 - y_2))));
}

auto main() -> int {
  std::vector<std::array<double, 2>> points;

  std::ifstream pla85900("../../../pla85900.tsp");
  std::string line;

  while (std::getline(pla85900, line)) {
    std::istringstream lineStream(line);

    double x_coord;
    double y_coord;

    if (int index; lineStream >> index >> x_coord >> y_coord) {
      points.push_back({x_coord, y_coord});
    }
  }

  pla85900.close();

  auto start = std::chrono::high_resolution_clock::now();

  sort_2d(points);

  auto stop = std::chrono::high_resolution_clock::now();

  auto duration =
      std::chrono::duration_cast<std::chrono::milliseconds>(stop - start)
          .count();

  double dist = 0;

  for (size_t i = 0; i < points.size() - 1; i++) {
    dist += calc_dist(points[i], points[i + 1]);
  }

  dist += calc_dist(points[points.size() - 1], points[0]);

  std::cout << "time: " << duration << ".\n"
            << "dist: " << static_cast<unsigned>(dist) << ".\n"
            << "mult: " << std::fixed << dist / 142382641 << ".\n";

  // time: 35.
  // dist: 188465250.
  // mult: 1.323653.
  // 11th Gen Intel(R) Core(TM) i5-11320H @ 3.20GHz
}

File: benchmark/js/benchmark.js
import { readFile } from "node:fs/promises";
import { sort2D } from "./hilbert_curve_sort.js";

function calcDist([x1, y1], [x2, y2]) {
	return Math.ceil(Math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2));
}

const points = [];

const pla85900 = await readFile("pla85900.tsp", "ascii");
const lines = pla85900.split("\n");

for (let i = 6; i < lines.length - 1; i++) {
	points.push(lines[i].split(/\s+/).slice(1).map(Number));
}

const start = performance.now();

const sorted = sort2D(points);

const stop = performance.now();

const duration = stop - start;

let dist = 0;

for (let i = 0; i < sorted.length - 1; i++) {
	dist += calcDist(sorted[i], sorted[i + 1]);
}

dist += calcDist(sorted.at(-1), sorted[0]);

console.log("time:", duration);
console.log("dist:", dist);
console.log("mult:", (dist / 142382641).toFixed(6));

// time: 170.
// dist: 188465250.
// mult: 1.323653.
// 11th Gen Intel(R) Core(TM) i5-11320H @ 3.20GHz
